### ISSUE TITLE
Use `electron.net` in the `github-authentication` extension

### ISSUE THREAD

**Issue by devm33** [microsoft/vscode#207867]
With https://github.com/microsoft/vscode/issues/192899 and https://github.com/microsoft/vscode/pull/206822 extensions have the ability to use [`electron.net.fetch`](https://www.electronjs.org/docs/latest/api/net) to benefit from improved proxy and networking capabilities offered by Electron and Chromium.

For `github-authentication`, this could replace the use of [`node-fetch`](https://github.com/microsoft/vscode/blob/b9b93f2b45aa90e4d5f3b80351145d6f61203922/extensions/github-authentication/src/node/fetch.ts).

**Comment by TylerLeonhardt** [comment]
@deepak1556 @alexdima @chrmarti do either of you know what work is involved for me to adopt this?

**Comment by deepak1556** [comment]
Fetch API in Node.js and Electron share the same API design, so you should be able to just swap in the imports and everything should just work. The minimal changes needed would be

```
import net from 'electron';

export const fetching = net.fetch;
```

However, I would recommend to try this out as experiment similar to how copilot extension does it, refs https://github.com/microsoft/vscode-copilot/pull/3461

**Comment by devm33** [comment]
@TylerLeonhardt I noticed https://github.com/microsoft/vscode/pull/229202 going in and just wanted make sure github-authentication doesn't get left behind 😄 I don't think it will pick up https://github.com/microsoft/vscode/pull/228476 currently since it's using `node-fetch` https://github.com/microsoft/vscode/blob/f26394d47799728503e3107f4807cf846966952c/extensions/github-authentication/src/node/fetch.ts#L5-L7

**Comment by chrmarti** [comment]
Note that I will change the default for https://github.com/microsoft/vscode/pull/229202 to be off for the stable release. Since this affects all extensions using `fetch` we first want to test it more in next milestone's Insiders.

**Comment by TylerLeonhardt** [comment]
@devm33 do you have any way of confirming that this is now working for folks in Insiders?

**Comment by devm33** [comment]
> [@devm33](https://github.com/devm33) do you have any way of confirming that this is now working for folks in Insiders?

@TylerLeonhardt I think I'm seeing an issue testing on latest insiders:

```
Version: 1.97.0-insider (Universal)
Commit: a1fc8c144985285527fcceb7adfa5f66b6bb5399
Date: 2025-01-27T05:04:55.863Z
Electron: 32.2.7
ElectronBuildId: 10660205
Chromium: 128.0.6613.186
Node.js: 20.18.1
V8: 12.8.374.38-electron.0
OS: Darwin arm64 23.6.0
```

```
2025-01-28 02:42:27.506 [info] Got 0 sessions for user:email...
2025-01-28 02:43:35.995 [info] Logging in with 'any' account...
2025-01-28 02:43:35.995 [info] Logging in for the following scopes: user:email
2025-01-28 02:43:36.004 [info] Trying without local server... (user:email)
2025-01-28 02:43:39.911 [info] Exchanging code for token...
2025-01-28 02:43:39.930 [error] fetch failed
2025-01-28 02:44:11.828 [info] Trying with local server... (user:email)
2025-01-28 02:44:20.666 [info] Exchanging code for token...
2025-01-28 02:44:20.699 [error] fetch failed
2025-01-28 02:47:27.204 [info] Logging in with 'any' account...
2025-01-28 02:47:27.204 [info] Logging in for the following scopes: user:email
2025-01-28 02:47:27.207 [info] Trying without local server... (user:email)
2025-01-28 02:47:32.387 [info] Exchanging code for token...
2025-01-28 02:47:32.408 [error] fetch failed
```

Is there a way to get more verbose logging or see if any error was thrown?

It looks like it's failing on this fetch: https://github.com/microsoft/vscode/blob/7a3d738bbb52a0222cbd97277b07d93113b79139/extensions/github-authentication/src/flows.ts#L103

But I don't see where the `[error] fetch failed` is coming from in the extension?

**Comment by chrmarti** [comment]
I have seen `fetch failed` as the only error message with Node.js' `fetch`. Could you check the output of `F1` > `Developer: GitHub Copilot Chat Diagnostics`?  Are you using a proxy? Or a Remote Development extension?

**Comment by TylerLeonhardt** [comment]
> The github-authentication extension runs in the standard extension host right?

Yes, nothing special there. It runs on the client's extension host

**Comment by chrmarti** [comment]
I'll open a new issue for Node.js' fetch failing. Not sure why the above code snippets wouldn't do the same, maybe something with module loading? Best to step through with the debugger.

**Comment by devm33** [comment]
Should 6263fffe0edbb0548e5c296a6760d34d72711856 be in the latest insiders? Looking at `/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/extensions/github-authentication/dist/extension.js` I don't see any `require("electron")`

Is the github-authentication published from the same commit as the main app?

Version: 1.97.0-insider (Universal)
Commit: a1fc8c144985285527fcceb7adfa5f66b6bb5399
Date: 2025-01-27T05:04:55.863Z
Electron: 32.2.7
ElectronBuildId: 10660205
Chromium: 128.0.6613.186
Node.js: 20.18.1
V8: 12.8.374.38-electron.0
OS: Darwin arm64 23.6.0

**Comment by TylerLeonhardt** [comment]
It's there I'm pretty sure:
```
"use strict";let r;Object.defineProperty(t,"__esModule",{value:!0}),t.fetching=void 0;try{r=n(2556).net.fetch}catch{r=fetch}t.fetching=r}
```
`n(2556)` probably maps to `require('electron')`

**Comment by devm33** [comment]
Thanks @TylerLeonhardt I think I see the issue. In the bundled output, `2556` is mapped to:

```js
2556: (e, t, n) => {
        const r = n(9896),
          i = n(6928),
          o = i.join(__dirname, "path.txt");
        e.exports = (function () {
          let e;
          if (
            (r.existsSync(o) && (e = r.readFileSync(o, "utf-8")),
            process.env.ELECTRON_OVERRIDE_DIST_PATH)
          )
            return i.join(
              process.env.ELECTRON_OVERRIDE_DIST_PATH,
              e || "electron"
            );
          if (e) return i.join(__dirname, "dist", e);
          throw new Error(
            "Electron failed to install correctly, please delete node_modules/electron and try installing again"
          );
        })();
```

which is the https://www.npmjs.com/package/electron package. I think we need to add `electron` to [the webpack config](https://github.com/microsoft/vscode/blob/724d644bbdb64a2d15c2ccdaa6d1ae9b23b6b0e4/extensions/github-authentication/extension.webpack.config.js) `externals`.

Though notably that's not what [extensions/microsoft-authentication/extension.webpack.config.js](https://github.com/microsoft/vscode/blob/724d644bbdb64a2d15c2ccdaa6d1ae9b23b6b0e4/extensions/microsoft-authentication/extension.webpack.config.js) has so I'm not sure how (if?) that's working.

**Comment by TylerLeonhardt** [comment]
Keep in mind, Microsoft Auth has a new implementation that doesn't use `'electron'` unless you explicitly ask for the `classic` version of the extension. In my testing, this new implementation provided by MSAL-node was quite thorough with proxy stuff.

**Comment by devm33** [comment]
Updated [c4b0c17](https://github.com/microsoft/vscode/pull/239134/commits/c4b0c173773d44283788f0aea9ddb85efa94a462)

**Comment by TylerLeonhardt** [comment]
Fixed in https://github.com/microsoft/vscode/pull/239134

**Comment by TylerLeonhardt** [comment]
@devm33 verified this one
